### PR TITLE
Handle error detail client error responses

### PR DESCRIPTION
**Pull request by thanhtrantruong47**
Handle error detail when respone status 404

image

### CODE REVIEW

File: src/hooks/useProductById.ts [hunk 1]
```diff
@@ -8,31 +8,42 @@ const useMockData = import.meta.env.USE_MOCK_FOR_API_FAIL === 'true';
...
   product: ProductType | null;
   isLoading: boolean;
+  status?: number; // Optional status code
 }
 
-const fetchProductById = async (productId: string) => {
+const fetchProductById = async (productId: string, useMockData: boolean) => {
   if (!productId) {
-    return null;
+    return { data: null, status: 0 };
```
Comment (hungdo-agilityio) [inline src/hooks/useProductById.ts]: `status: 404` instead you can read more at https://developer.mozilla.org/en-US/docs/Web/HTTP/Status

File: src/hooks/useProductById.ts [hunk 2]
```diff
@@ -8,31 +8,42 @@ const useMockData = import.meta.env.USE_MOCK_FOR_API_FAIL === 'true';
...
   }
 
-  const response = await ProductAPIService.getById(productId);
-
-  if (response.isSuccess && response.data !== undefined && !useMockData) {
-    return response.data;
+  if (useMockData) {
+    const mockProduct = mockProducts.find(product => product.id.toString() === productId) || null;
+    return { data: mockProduct, status: 200 }; // Assuming mock data is always successful
```
Comment (hungdo-agilityio) [inline src/hooks/useProductById.ts:23]: add break line before return, recheck for all

File: src/hooks/useProductById.ts [hunk 3]
```diff
@@ -8,31 +8,42 @@ const useMockData = import.meta.env.USE_MOCK_FOR_API_FAIL === 'true';
...
+  } catch (error) {
+    console.error('Failed to fetch product:', error);
+    return { data: null, status: 500 }; // Assuming a 500 status code for errors
+  }
 };
 
 export const useProductById = (id: string): UseProductByIdResult => {
   const [product, setProduct] = useState<ProductType | null>(null);
   const [isLoading, setIsLoading] = useState(false);
+  const [status, setStatus] = useState<number | undefined>(undefined);
```
Comment (hungdo-agilityio) [inline src/hooks/useProductById.ts]: `const [status, setStatus] = useState<number>();`

File: src/pages/ProductDetail/index.tsx [hunk 4]
```diff
@@ -9,41 +9,58 @@ import { Product, Loading, ProductList } from '@/components';
...
 
+  // Use the useProductById hook for the product and status
+  const { product, isLoading: isProductLoading, status } = useProductById(id || '');
+
   // Use the useProducts hook for related products
-  const { product, isLoading: isProductLoading } = useProductById(id || '');
   const { products, isLoading: isRelatedProductsLoading, hasMore, onLoadMore } = useProducts(product?.category || '');
 
+  // Determine loading state
+  const isLoading = isProductLoading && status === undefined;
```
Comment (hungdo-agilityio) [inline src/pages/ProductDetail/index.tsx]: Do we need to check status for flag isLoading?

File: src/pages/ProductDetail/index.tsx [hunk 5]
```diff
@@ -9,41 +9,58 @@ import { Product, Loading, ProductList } from '@/components';
...
-                rate={product.rate || 0}
-                sizes={product.sizes || []}
-                stock={product.stock || 0}
-                title={product.title || ''}
-              />
-            )
-          )}
+      {isLoading ? (
+        <Loading classStyle={`${styleUtils.loading} ${styles.loading}`} />
+      ) : status === 404 ? (
```
Comment (hungdo-agilityio) [inline src/pages/ProductDetail/index.tsx]: create variable for this `const isNotFound = status === STATUS.NOT_FOUND`

File: src/pages/ProductDetail/index.tsx [hunk 6]
```diff
@@ -9,41 +9,58 @@ import { Product, Loading, ProductList } from '@/components';
...
-          <p className={styles.title}>Same Product</p>
-          {isRelatedProductsLoading && products.length === 0 ? (
-            <Loading classStyle={styleUtils.loading} />
-          ) : (
-            <ProductList products={products} onClick={onLoadMore} hasMore={hasMore} />
-          )}
+      ) : (
+        <div>
+          <div className={styles.product}>
+            {product ? (
```
Comment (hungdo-agilityio) [inline src/pages/ProductDetail/index.tsx]: No need to check if product or not here, include this in isNotFound above
`const isNotFound = status === STATUS.NOT_FOUND || !product`

File: src/hooks/useProductById.ts (hunk 1 above)
Comment (thanhtrantruong47) [reply src/hooks/useProductById.ts]: Fixed

File: src/hooks/useProductById.ts (hunk 2 above)
Comment (thanhtrantruong47) [reply src/hooks/useProductById.ts:23]: Fixed

File: src/pages/ProductDetail/index.tsx (hunk 5 above)
Comment (thanhtrantruong47) [reply src/pages/ProductDetail/index.tsx]: Fixed

File: src/hooks/useProductById.ts (hunk 3 above)
Comment (thanhtrantruong47) [reply src/hooks/useProductById.ts]: Fixed

File: src/pages/ProductDetail/index.tsx (hunk 4 above)
Comment (thanhtrantruong47) [reply src/pages/ProductDetail/index.tsx]: Fixed

File: src/pages/ProductDetail/index.tsx (hunk 6 above)
Comment (thanhtrantruong47) [reply src/pages/ProductDetail/index.tsx]: Fixed

File: src/constants/index.ts [hunk 7]
```diff
@@ -57,3 +57,7 @@ export enum MESSAGE_SUCCESS {
   ADD_TO_CART = 'Item added to cart!',
   DELETE_CART = 'Delete item in cart',
 }
+
+export enum STATUS {
+  NOT_FOUND = 404,
+}
```
Comment (hungdo-agilityio) [inline src/constants/index.ts:63]: Please add more status for example 500, I saw some place use it

**Review by hungdo-agilityio — APPROVED** [review]

**Pull request merged**